### PR TITLE
Testing composite index queries against production

### PR DESCRIPTION
**Pull request by milaGGL**
- [x] create a `CompositeIndexTestHelper` class to run tests with composite index against production
- [x] add terraform to create composite indexes inside `packages/firestore`
- [x] move part of the OR queries to composite index query test file.

### CODE REVIEW

File: .github/workflows/test-changed-firestore.yml [hunk 1]
```diff
@@ -118,6 +118,8 @@ jobs:
         run: cp config/ci.config.json config/project.json
       - name: Run tests
         run: cd packages/firestore && yarn run ${{ matrix.test-name }}
+        env:
```
Comment (milaGGL) [inline .github/workflows/test-changed-firestore.yml:121]: code from other PR.

File: scripts/emulator-testing/emulators/firestore-emulator.ts [hunk 2]
```diff
@@ -22,11 +22,11 @@ export class FirestoreEmulator extends Emulator {
 
   constructor(port: number, projectId = 'test-emulator') {
     super(
-      'cloud-firestore-emulator-v1.14.4.jar',
```
Comment (milaGGL) [inline scripts/emulator-testing/emulators/firestore-emulator.ts:25]: code from other PR.

File: main.tf [hunk 3]
```diff
@@ -0,0 +1,39 @@
+provider "google" {
+  project = local.project.projectId
+}
+
+resource "google_firestore_index" "default-db-index" {
+  collection = "composite-index-test-collection"
+
+  for_each = local.indexes
+  dynamic "fields" {
```
Comment (milaGGL) [inline main.tf:9]: Seperated the indexes out into `firestore_index_config.tf` file for development simplicity.

File: main.tf [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      field_path = lookup(fields.value, "field_path", null)
+      order      = lookup(fields.value, "order", null)
+    }
+  }
+
+}
+
+resource "google_firestore_index" "named-db-index" {
+  collection = "composite-index-test-collection"
+  database   = "test-db"
```
Comment (milaGGL) [inline main.tf:25]: creates indexes for "test-db" database for named-DB tests.

File: packages/firestore/test/integration/util/composite_index_test_helper.ts [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+  ): Query<AppModelType, DbModelType> {
+    return internalQuery(
+      query_,
+      and(where(this.TEST_ID_FIELD, '==', this.testId), compositeFilter),
+      ...queryConstraints
+    );
+  }
+
+  // Add a document with test id.
+  addDoc<T, DbModelType extends DocumentData>(
```
Comment (milaGGL) [inline packages/firestore/test/integration/util/composite_index_test_helper.ts:147]: addDoc and setDoc is not used in current test cases. Should I keep these for future use or remove it as it is not required for now?

**Review by ehsannas — COMMENTED** [review]
Generally great! Just a couple of minor nits pointed out below.

File: main.tf (hunk 4 above)
Comment (milaGGL) [reply main.tf:25]: It is default to "(default)" DB.

File: packages/firestore/test/integration/util/composite_index_test_helper.ts [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+  // Runs a test without pre-created documents in the COMPOSITE_INDEX_TEST_COLLECTION.
+  async withEmptyCollection<T>(
+    persistence: PersistenceMode | typeof PERSISTENCE_MODE_UNSPECIFIED,
+    fn: (collection: CollectionReference, db: Firestore) => Promise<T>
+  ): Promise<T> {
+    return batchCommitDocsToCollection(
+      persistence,
+      DEFAULT_SETTINGS,
+      {},
```
[the remark this answers is not in the document]
Comment (milaGGL) [reply packages/firestore/test/integration/util/composite_index_test_helper.ts:85]: it is. 

I added emptyCollection on for convenience. But after re-thinking it, it is not so necessary. I will remove it.

**Review by ehsannas — APPROVED** [review]

File: .github/workflows/test-changed-firestore-integration.yml [hunk 7]
```diff
@@ -13,6 +13,21 @@ jobs:
...
+    # create composite indexes with Terraform
+    - name: Setup Terraform 
+      uses: hashicorp/setup-terraform@v2
+    - name: Terraform Init
+      run: terraform init
+      continue-on-error: true
+    - name: Terraform Apply
+      if: github.event_name == 'pull_request'
+      run: |
+        cp config/ci.config.json config/project.json
```
Comment (DellaBitta) [inline .github/workflows/test-changed-firestore-integration.yml:28]: This happens below, too. Might be better to move the step on line 43 up above this step and then only have the `terraform apply` in this step.

Comment (milaGGL) [reply .github/workflows/test-changed-firestore-integration.yml:28]: Done.

File: packages/firestore/test/integration/util/composite_index_test_helper.ts [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+ *
+ * Key Features:
+ * - Runs tests against the dedicated test collection with predefined composite indexes.
+ * - Automatically associates a test ID with documents for data isolation.
+ * - Utilizes TTL policy for automatic test data cleanup.
+ * - Constructs Firestore queries with test ID filters.
+ */
+export class CompositeIndexTestHelper {
+  private readonly testId: string;
+  private readonly TEST_ID_FIELD: string = 'testId';
```
Comment (ehsannas) [inline packages/firestore/test/integration/util/composite_index_test_helper.ts:57]: please add

```js
  private readonly TTL_FIELD: string = 'expireAt';
```

and use `TTL_FIELD` on line 94

File: packages/firestore/test/integration/util/composite_index_test_helper.ts [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+  // Sets a document in Firestore with test-specific fields.
+  setDoc<T, DbModelType extends DocumentData>(
+    reference: DocumentReference<T, DbModelType>,
+    data: object
+  ): Promise<void> {
+    const processedData = this.addTestSpecificFieldsToDoc(
+      data
+    ) as WithFieldValue<T>;
+    return setDocument(reference, processedData);
+  }
```
Comment (ehsannas) [inline packages/firestore/test/integration/util/composite_index_test_helper.ts:174]: I know these functions aren't needed right now, but we should probably implement them to have a complete set of utility functions in this class. (_note: this is rough untested code._)

```js
// Updates a document in Firestore.
// This is is the same as making the update on the doc directly with merge=true.
updateDoc<T, DbModelType extends DocumentData>(
    reference: DocumentReference<T, DbModelType>,
    updates: object
  ): Promise<void> {
  return updateDoc(reference, updates);
}

// Gets the document content.
async getDoc<AppModelType, DbModelType extends DocumentData>(
    reference: DocumentReference<AppModelType, DbModelType>
): Promise<DocumentSnapshot<AppModelType, DbModelType>> {
    let docSnapshot = await getDocument<AppModelType, DbModelType>(reference);
    docSnapshot._document?.data?.delete(field(this.TTL_FIELD));
    docSnapshot._document?.data?.delete(field(this.TEST_ID_FIELD));
    return docSnapshot;
}

getDocs<AppModelType, DbModelType extends DocumentData>(
  query: Query<AppModelType, DbModelType>
): Promise<QuerySnapshot<AppModelType, DbModelType>> {
  // querySnapshot = await getDocs(query);
  // iterate over documents in the querySnapshot, and do the same thing as above.
  // should probably make a function named "removeTestSpecificFieldsFromDoc" and
  // and call that in these two functions.
}
```

**Review by ehsannas — COMMENTED** [review]
One final request below.

File: packages/firestore/test/integration/util/composite_index_test_helper.ts [hunk 10]
```diff
@@ -172,4 +187,34 @@ export class CompositeIndexTestHelper {
     ) as WithFieldValue<T>;
     return setDocument(reference, processedData);
   }
+
+  // This is is the same as making the update on the doc directly with merge=true.
```
Comment (ehsannas) [inline packages/firestore/test/integration/util/composite_index_test_helper.ts]: typo: `this is is` --> `This is`

**Review by DellaBitta — APPROVED** [review]

**Pull request merged**